File: backend/app/intelligence/data_quality.py
from typing import Dict, Any, List

class DataQualityEngine:
    SECTOR_RANGES = {
        "cement": {"gei": (0.45, 1.20), "elec_kwh_per_t": (50, 140), "clinker_factor": (45.0, 98.0)},
        "aluminium": {"gei": (10.0, 20.0), "elec_kwh_per_t": (12000, 17000)},
        "chlor_alkali": {"gei": (1.20, 2.50), "elec_kwh_per_t": (2000, 3200)},
        "pulp_paper": {"gei": (1.20, 3.20), "elec_kwh_per_t": (600, 1400)},
        "petrochemicals": {"gei": (0.70, 2.20), "elec_kwh_per_t": (400, 1100)},
        "petroleum_refinery": {"gei": (0.15, 0.50), "elec_kwh_per_t": (40, 120)},
        "iron_steel": {"gei": (1.60, 3.40), "elec_kwh_per_t": (400, 1200)},
        "textile": {"gei": (2.50, 7.50), "elec_kwh_per_t": (1000, 3500)},
        "fertiliser": {"gei": (0.35, 0.95), "elec_kwh_per_t": (80, 250)},
    }

    @staticmethod
    def audit_facility_input(data: Dict[str, Any]) -> Dict[str, Any]:
        errors = []
        warnings = []
        info = []
        score = 100.0

        sector = data.get("sector", "").lower()
        output = data.get("annual_production", 0.0)
        elec_mwh = data.get("electricity_mwh", 0.0)
        renewable_pct = data.get("renewable_electricity_pct", 0.0)
        fuel_qty = data.get("thermal_fuel_tonnes", 0.0)
        fuel_type = data.get("thermal_fuel_type", "coal")

        # 1. Mandatory & Non-Negative Checks
        if not sector:
            errors.append("Sector is required.")
            score -= 40.0
        if output <= 0:
            errors.append("Annual production output must be greater than zero.")
            score -= 40.0

        if elec_mwh < 0:
            errors.append("Electricity consumption cannot be negative.")
            score -= 20.0
        if fuel_qty < 0:
            errors.append("Fuel quantity cannot be negative.")
            score -= 20.0
        if renewable_pct < 0 or renewable_pct > 100:
            errors.append("Renewable electricity percentage must be between 0% and 100%.")
            score -= 15.0

        # 2. Activity Consistency Checks
        if output > 0 and elec_mwh == 0 and fuel_qty == 0:
            errors.append("Both electricity and thermal fuel quantities are zero for operational facility.")
            score -= 35.0

        # 3. Sector Thermodynamic Sanity Ranges
        if sector in DataQualityEngine.SECTOR_RANGES and output > 0:
            ranges = DataQualityEngine.SECTOR_RANGES[sector]
            elec_intensity_kwh = (elec_mwh * 1000.0) / output if output > 0 else 0

            if "elec_kwh_per_t" in ranges:
                min_e, max_e = ranges["elec_kwh_per_t"]
                if elec_intensity_kwh < min_e * 0.5:
                    warnings.append(f"Electricity intensity ({elec_intensity_kwh:.1f} kWh/unit) is unusually low for {sector.capitalize()} (typical: {min_e}-{max_e} kWh/unit).")
                    score -= 10.0
                elif elec_intensity_kwh > max_e * 2.0:
                    warnings.append(f"Electricity intensity ({elec_intensity_kwh:.1f} kWh/unit) is significantly higher than industry benchmark (typical: {min_e}-{max_e} kWh/unit).")
                    score -= 10.0

        # 4. Sector-Specific Process Checks
        if sector == "cement":
            cf = data.get("clinker_factor_pct", 72.0)
            if cf < 40.0 or cf > 98.0:
                warnings.append(f"Reported clinker factor ({cf:.1f}%) is outside typical industrial range (45% - 98%).")
                score -= 8.0

        status = "PASS" if not errors and score >= 80.0 else ("WARNING" if not errors else "BLOCKING_ERROR")
        final_score = max(0.0, min(100.0, round(score, 1)))

        return {
            "status": status,
            "quality_score": final_score,
            "errors": errors,
            "warnings": warnings,
            "info": info,
            "is_valid": len(errors) == 0,
            "data_category": "USER_SUBMITTED_VALIDATED"
        }

File: lib/engines/dataQuality.ts
/**
 * AANGARA Data Quality & MRV Verification Engine v3.0
 * ─────────────────────────────────────────────────────────
 * Implements:
 *   §34.4: Weighted Completeness Model
 *   §34.5: Ledger Reconciliation & Invariant Checks
 *   §34.6: Six Independent Confidence Dimensions
 *   §34.7: Five-Level Decision Readiness Gate
 *
 * Fully backward-compatible with v1/v2 audit calls while providing
 * comprehensive audit trace for ACVA statutory verification readiness.
 */

export interface CompletenessItem {
  key: string;
  label: string;
  weight: number;
  available: boolean;
  source_type?: string;
  data_class?: string;
}

export interface ConfidenceDimension {
  dimension: string;
  score: number; // 0–100
  tier: 'HIGH' | 'MEDIUM' | 'LOW' | 'PROVISIONAL';
  description: string;
}

export type DecisionReadinessLevel =
  | 'DATA_READY'           // Level 1: Minimal fields present
  | 'CALCULATION_READY'    // Level 2: Physical energy & mass balances closed
  | 'BENCHMARK_READY'      // Level 3: Verified peer cohort mapping possible
  | 'PROJECT_MODEL_READY'  // Level 4: Abatement engineering costed
  | 'DECISION_READY';      // Level 5: Audit-ready for Board sign-off

export interface DataQualityAuditResult {
  // Backward compatibility fields
  is_valid: boolean;
  quality_score: number;
  quality_tier: 'HIGH_CONFIDENCE' | 'PROVISIONAL' | 'ACTION_REQUIRED';
  errors: { field: string; message: string }[];
  warnings: { field: string; message: string }[];

  // v3.0 Enriched Fields
  completeness_score: number;
  reconciliation_errors: {
    electricity_reconciliation_pct: number;
    fuel_reconciliation_pct: number;
    gei_recomputed_delta_pct: number;
  };
  confidence_dimensions: {
    regulatory_evidence_confidence: ConfidenceDimension;
    operational_data_confidence: ConfidenceDimension;
    energy_data_confidence: ConfidenceDimension;
    factor_confidence: ConfidenceDimension;
    project_cost_confidence: ConfidenceDimension;
    market_input_confidence: ConfidenceDimension;
  };
  readiness_level: DecisionReadinessLevel;
  readiness_description: string;
  acva_verification_readiness_pct: number;
}

export class DataQualityEngine {

  /**
   * Evaluates input completeness across 5 key information categories (§34.4).
   */
  private static evaluateCompleteness(data: Record<string, any>): {
    score: number;
    items: CompletenessItem[];
  } {
    const prod = Number(data.annual_production ?? data.production?.reporting_year_production ?? 0);
    const elec = Number(data.electricity_mwh ?? 0);
    const has_elec_ledger = Array.isArray(data.carbon_inputs?.electricity_sources) && data.carbon_inputs.electricity_sources.length > 0;
    const has_fuel_ledger = Array.isArray(data.carbon_inputs?.fuel_streams) && data.carbon_inputs.fuel_streams.length > 0;
    const has_fuel_scalar = Number(data.thermal_fuel_tonnes ?? 0) > 0;
    const sector = data.sector ?? data.regulatory?.sector;
    const has_process = !!(data.process_inputs || data.clinker_factor_pct || data.smelter_dc_sec_kwh);
    const has_finance = !!(data.finance_inputs || data.primary_project || data.capex_cr);

    const items: CompletenessItem[] = [
      { key: 'production', label: 'Production / Activity Output', weight: 25, available: prod > 0 },
      { key: 'electricity', label: 'Scope 2 Electricity Data', weight: 25, available: elec > 0 || has_elec_ledger },
      { key: 'thermal_fuels', label: 'Scope 1 Thermal Fuel Data', weight: 25, available: has_fuel_scalar || has_fuel_ledger },
      { key: 'process_data', label: 'Sector Process Parameters', weight: 15, available: !!sector && has_process },
      { key: 'financial_data', label: 'Financial & Capital Parameters', weight: 10, available: has_finance },
    ];

    const totalWeight = items.reduce((s, it) => s + it.weight, 0);
    const earnedWeight = items.filter((it) => it.available).reduce((s, it) => s + it.weight, 0);
    const score = Number(((earnedWeight / Math.max(1, totalWeight)) * 100).toFixed(1));

    return { score, items };
  }

  /**
   * Runs reconciliation between summary numbers and ledger stream sums (§34.5).
   */
  private static evaluateReconciliations(data: Record<string, any>) {
    let elec_reconcil_pct = 0;
    let fuel_reconcil_pct = 0;
    let gei_delta_pct = 0;

    // Electricity check
    if (data.carbon_inputs?.electricity_sources) {
      const sumSources = (data.carbon_inputs.electricity_sources as any[]).reduce(
        (s, r) => s + (Number(r.annual_mwh) || 0),
        0
      );
      const reported = Number(data.electricity_mwh || sumSources);
      if (reported > 0) {
        elec_reconcil_pct = Number((Math.abs(reported - sumSources) / reported * 100).toFixed(2));
      }
    }

    // Fuel check
    if (data.carbon_inputs?.fuel_streams) {
      const sumStreams = (data.carbon_inputs.fuel_streams as any[]).reduce(
        (s, r) => s + (Number(r.quantity) || 0),
        0
      );
      const reported = Number(data.thermal_fuel_tonnes || sumStreams);
      if (reported > 0) {
        fuel_reconcil_pct = Number((Math.abs(reported - sumStreams) / reported * 100).toFixed(2));
      }
    }

    return {
      electricity_reconciliation_pct: elec_reconcil_pct,
      fuel_reconciliation_pct: fuel_reconcil_pct,
      gei_recomputed_delta_pct: gei_delta_pct,
    };
  }

  /**
   * Main Facility Data Audit Method
   */
  static auditFacilityInput(data: Record<string, any>): DataQualityAuditResult {
    const errors: { field: string; message: string }[] = [];
    const warnings: { field: string; message: string }[] = [];

    // ── 1. Validation Invariants ──
    const prod = Number(data.annual_production ?? data.production?.reporting_year_production ?? 0);
    if (prod <= 0) {
      errors.push({ field: 'annual_production', message: 'Annual production must be greater than 0.' });
    }

    const elec = Number(data.electricity_mwh ?? 0);
    if (elec < 0) {
      errors.push({ field: 'electricity_mwh', message: 'Electricity consumption cannot be negative.' });
    }

    const ren_pct = Number(data.renewable_electricity_pct ?? 0);
    if (ren_pct < 0 || ren_pct > 100) {
      errors.push({ field: 'renewable_electricity_pct', message: 'Renewable electricity percentage must be between 0 and 100%.' });
    }

    // Sector-specific engineering reasonableness checks
    const sec = (data.sector ?? data.regulatory?.sector ?? '').toLowerCase();
    if (sec === 'cement') {
      const elec_int = prod > 0 ? (elec * 1000) / prod : 0;
      if (elec > 0 && (elec_int < 50 || elec_int > 160)) {
        warnings.push({
          field: 'electricity_mwh',
          message: `Specific electrical consumption (${elec_int.toFixed(1)} kWh/t) outside typical Indian cement range (65-115 kWh/t).`,
        });
      }
    } else if (sec === 'aluminium') {
      const elec_int = prod > 0 ? (elec * 1000) / prod : 0;
      if (elec > 0 && (elec_int < 12000 || elec_int > 16500)) {
        warnings.push({
          field: 'electricity_mwh',
          message: `Smelter DC electrical consumption (${elec_int.toFixed(0)} kWh/t) outside standard Hall-Héroult range (13,000–15,500 kWh/t).`,
        });
      }
    }

    // ── 2. Completeness Evaluation (§34.4) ──
    const completeness = this.evaluateCompleteness(data);

    // ── 3. Ledger Reconciliations (§34.5) ──
    const reconciliations = this.evaluateReconciliations(data);
    if (reconciliations.electricity_reconciliation_pct > 5.0) {
      warnings.push({
        field: 'electricity_ledger',
        message: `Reported electricity differs by ${reconciliations.electricity_reconciliation_pct}% from sum of source ledgers.`,
      });
    }

    if (reconciliations.fuel_reconciliation_pct > 5.0) {
      warnings.push({
        field: 'fuel_ledger',
        message: `Reported thermal fuel differs by ${reconciliations.fuel_reconciliation_pct}% from sum of fuel streams.`,
      });
    }

    // ── 4. Quality Scoring Formula (§34.4) ──
    let score = completeness.score;
    score -= errors.length * 25.0;
    score -= warnings.length * 6.0;
    score -= (reconciliations.electricity_reconciliation_pct > 1.0 ? 5.0 : 0);
    score -= (reconciliations.fuel_reconciliation_pct > 1.0 ? 5.0 : 0);
    score = Math.max(0.0, Math.min(100.0, Number(score.toFixed(1))));

    const quality_tier = score >= 85 ? 'HIGH_CONFIDENCE' : (score >= 60 ? 'PROVISIONAL' : 'ACTION_REQUIRED');

    // ── 5. Six Confidence Dimensions (§34.6) ──
    const makeDim = (
      name: string,
      scoreVal: number,
      desc: string
    ): ConfidenceDimension => ({
      dimension: name,
      score: scoreVal,
      tier: scoreVal >= 80 ? 'HIGH' : scoreVal >= 60 ? 'MEDIUM' : 'LOW',
      description: desc,
    });

    const confidence_dimensions = {
      regulatory_evidence_confidence: makeDim(
        'Regulatory Evidence',
        sec && ['cement', 'aluminium', 'chlor_alkali', 'pulp_paper', 'petrochemicals', 'petroleum_refinery', 'textile'].includes(sec) ? 90 : 50,
        'Gazette statutory baseline & target coverage under CCTS notifications.'
      ),
      operational_data_confidence: makeDim(
        'Operational Data',
        prod > 0 ? 85 : 30,
        'Production and capacity utilization data traceability.'
      ),
      energy_data_confidence: makeDim(
        'Energy Data',
        elec > 0 && reconciliations.electricity_reconciliation_pct <= 2.0 ? 88 : 65,
        'Electricity & thermal fuel metering closure with grid billing evidence.'
      ),
      factor_confidence: makeDim(
        'Emission Factor Quality',
        85,
        'CEA Baseline Database v21.0 & IPCC Tier 2 default factors applied via factor registry.'
      ),
      project_cost_confidence: makeDim(
        'Project Cost Confidence',
        data.primary_project || data.capex_cr ? 75 : 45,
        'Engineering CAPEX and energy saving estimates.'
      ),
      market_input_confidence: makeDim(
        'Market Inputs',
        70,
        'Scenario carbon pricing and financing interest rate benchmarks.'
      ),
    };

    // ── 6. Five-Level Decision Readiness Gate (§34.7) ──
    let readiness_level: DecisionReadinessLevel = 'DATA_READY';
    let readiness_desc = 'Initial input data entered.';

    if (errors.length === 0 && prod > 0) {
      readiness_level = 'CALCULATION_READY';
      readiness_desc = 'Physical energy & emissions mass balance calculations validated.';

      if (sec && !['iron_steel'].includes(sec)) {
        readiness_level = 'BENCHMARK_READY';
        readiness_desc = 'Sector gazette targets verified and mapped.';

        if (completeness.items.find((i) => i.key === 'financial_data')?.available) {
          readiness_level = 'PROJECT_MODEL_READY';
          readiness_desc = 'Decarbonisation capital intervention costed with cashflow profile.';

          if (score >= 80) {
            readiness_level = 'DECISION_READY';
            readiness_desc = 'Audit-grade verified twin ready for ACVA and executive decision.';
          }
        }
      }
    }

    const acva_verification_readiness_pct = Number(
      (0.35 * confidence_dimensions.energy_data_confidence.score +
       0.25 * confidence_dimensions.regulatory_evidence_confidence.score +
       0.20 * confidence_dimensions.factor_confidence.score +
       0.20 * confidence_dimensions.operational_data_confidence.score).toFixed(1)
    );

    return {
      is_valid: errors.length === 0,
      quality_score: score,
      quality_tier,
      errors,
      warnings,
      completeness_score: completeness.score,
      reconciliation_errors: reconciliations,
      confidence_dimensions,
      readiness_level,
      readiness_description: readiness_desc,
      acva_verification_readiness_pct,
    };
  }
}
